src/oop/programs_with_multiple_classes/hard/q4/MainTest.java

```diff
@@ -0,0 +1,150 @@
+     **/
+    @Test
+    public void bookClassHasCorrectFields() {
+        String missingFieldMessage = "Your " + bookLc + " class is missing a required field.";
+        String incorrectFieldModifierMessage = "A field in your " + bookLc + " class has an the wrong visibility modifier.";
+        assertTrue(book.hasField(year, int.class), missingFieldMessage);
+        assertTrue(book.hasField(type, String.class), missingFieldMessage);
+        assertTrue(book.hasModifier(year, "private"), incorrectFieldModifierMessage);
+        assertTrue(book.hasModifier(type, "private"), incorrectFieldModifierMessage);
+    }
+
+    @Test
+    public void bookClassHasCorrectConstructors() {
+        String missingConstructorMessage = "Your " + bookLc + " class does not have a required constructor.";
+        String incorrectConstructorModifierMessage =
+                "One of the constructors in your " + bookLc + " class does not have the correct visibility modifier.";
+        Class<?>[] constructorClasses = new Class[]{
+                int.class,
+                String.class
+        };
+        assertTrue(book.hasConstructor(constructorClasses), missingConstructorMessage);
+        assertTrue(book.hasModifier(constructorClasses, "public"), incorrectConstructorModifierMessage);
+    }
+
+    private static Stream<Arguments> bookInputProvider() {
+        return Stream.of(
+                Arguments.of(1, "Hardcover"),
+                Arguments.of(1, "Softcover"),
+                Arguments.of(3, "Paperback"),
+                Arguments.of(4, "Softback"),
+                Arguments.of(0, "Hardcover"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookInputProvider")
+    public void bookConstructorInitializesValuesCorrectly(int year, String type) throws Throwable {
+        Object[][] arguments = new Object[][]{
+                {year, int.class},
+                {type, String.class}
+        };
+        Object bookInstance = book.createInstance(arguments);
+        _assertEquals(year, book.getFieldValue(bookInstance, this.year),
+                "Your " + book + " constructor does not correctly initialize the " + this.year + " field.");
+        _assertEquals(type, book.getFieldValue(bookInstance, this.type),
+                "Your " + book + " constructor does not correctly initialize the " + this.type + " field.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookInputProvider")
+    public void correctHowManyYearsMethod(int year, String type) throws Throwable {
+        Object[][] arguments = new Object[][]{
+                {year, int.class},
+                {type, String.class}
+        };
+        Object bookInstance = book.createInstance(arguments);
+        Object howManyYearsOutput = book.callMethod(howManyYears, bookInstance);
+        int howManyYearsExpected = Year.now().getValue() - year;
+        String incorrectHowManyYearsExpected = "Your " + bookLc + " " + howManyYears + " method does not return the correct time.";
```

Comment:
I think it could be more helpful to the student if we said "... the correct number of years." instead of "... correct time". What do you think?